File: Pod/Classes/HSM/StateMachine.m
//
//  CPState.m
//  Model
//
//  Created by cnepayzx on 15-1-21.
//  Copyright (c) 2015年 cnepay. All rights reserved.
//

#import "StateMachine.h"
#import "State.h"
#import "CPMessagesQueue.h"
#import "CPMessageOperation.h"
#import "SmHandler.h"
#import "CPLoggerManager.h"
@interface StateMachine()

@property(nonatomic,strong) State *currentState;
@property (readwrite, nonatomic, strong) SmHandler *smHandler;
@property (readwrite, nonatomic, strong) QuittingState *quittingState;
@end

@implementation StateMachine

- (void) initialState:(State *)state {
    [self.smHandler setInitialState:state];
}
- (void)transitionToState:(State *)destState
{
    [self.smHandler transitionToState:destState];
}

+ (instancetype)stateMachine
{
    return [[self alloc] init];
}
- (void)quitNow
{
    [self.smHandler quitNow];
}

- (BOOL)isQuit:(CPMessage *)msg {
    
    if (self.smHandler) {
        return [self.smHandler isQuit:msg];
    }else {
        return msg.type = MessageTypeActionQuit;
    }
}

- (instancetype)init
{
    self = [super init];
    if (self) {
        _quittingState = [[QuittingState alloc] initWithStateMachine:self];
        _smHandler = [[SmHandler alloc] init];
        [_smHandler addState:_quittingState parentState:nil];
        _smHandler.handlerDelegate = self;
    }
    return self;
}

- (void)start
{
    CPStateMechineLog(@"self %@ %@",self,[NSThread currentThread]);
    [self.smHandler completeConstruction];
}


- (BOOL)hasDeferredMessage:(MessageType)type {
    for (CPMessage *msg in self.smHandler.deferredArray) {
        if (msg.type == type) {
            return YES;
        }
    }
    return NO;
    
}

- (void)removeDeferredMessage:(MessageType)type
{
    NSMutableArray *tempDeferredArray = [NSMutableArray array];
    for (CPMessage *msg in self.smHandler.deferredArray) {
        if (msg.type != type) {
            [tempDeferredArray addObject:msg];
        }
    }
    [self.smHandler.deferredArray removeAllObjects];
    self.smHandler.deferredArray = tempDeferredArray;
}

- (void)sendMessage:(CPMessage *)message
{
    [self.smHandler addOperationAtEndOfQueue:message];
}
- (void)sendMessageType:(MessageType)type
{
    [self.smHandler addOperationAtEndOfQueue:[CPMessage messageWithType:type]];
}

- (void)sendMessageFront:(CPMessage *)message
{
    [self.smHandler addOperationAtFrontOfQueue:message];
    
}
- (void)sendMessageFrontType:(MessageType)type
{
    
    [self.smHandler addOperationAtFrontOfQueue:[CPMessage messageWithType:type]];
}
- (void) removeMessageWithType:(MessageType)type {
    [self.smHandler.operations enumerateObjectsUsingBlock:^(__kindof CPMessageOperation * _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
        if (obj.message.type == type && ![obj isExecuting]) {
            [obj cancel];
        }
    }];
    
}


- (void)sendMessageDelayed:(CPMessage *)msg delay:(NSTimeInterval)delay
{
    [self.smHandler delayOperationAtEndOfQueue:msg delay:delay];
}

- (void)deferredMessage:(CPMessage *)message
{
    [self.smHandler.deferredArray addObject:message];
}

- (void)addState:(State *)state parentState:(State *)parentState
{
    [self.smHandler addState:state parentState:parentState];
}

- (void)smHandlerProcessFinalMessage:(CPMessage *)msg{}
@end

@implementation QuittingState
- (BOOL)processMessage:(CPMessage *)message
{
    return NO;
}
@end

File: Pod/Classes/HSM/StateMachine.swift
//
//  StateMachine.swift
//  StateMachine-Swift
//
//  Created by Cnepay on 15/11/2.
//  Copyright © 2015年 Cnepay. All rights reserved.
//

import UIKit

public class StateMachine: NSObject {
   public var runloopThread:NSThread?
//    static var unconnectState:State =  {
//        return UnconnectState()
//    }()
    
//    var destiedState:CPState?
//    var currentState:CPState!;
//    
//    
    var mSmHandler:SmHandler!;
    override init() {
        mSmHandler = SmHandler()
        
    }
    
    func start() {
        mSmHandler.completeConstruction()
    }
    
    func transitionTo(destState:State?) {
        self.mSmHandler.transitionTo(destState!);//.destState = destState
    }
//
//    
//    lazy var deferredArray:Array<CPMessage> = [CPMessage]()
//    
//    func transitionState(){
//        
//        while destiedState != nil && self.currentState !== destiedState {
//            self.currentState.exit();
//            self.currentState = self.destiedState!
//            self.currentState.enter()
//        }
//        self.destiedState = nil;
//    }
    
    func hasDeferredMessage(type:MESSAGETYPE) ->Bool {
        
        for message in self.mSmHandler.deferredArray {
            if message.type == type {
                return true
            }
        }
        return false;
    }
    
    func removeDeferredMessage(type:MESSAGETYPE) {
        var tempDeferredArray :Array<CPMessage> = [CPMessage]();
        
        for message in self.mSmHandler.deferredArray {
            if message.type != type {
                tempDeferredArray.append(message);
            }
        }
        self.mSmHandler.deferredArray.removeAll();
        self.mSmHandler.deferredArray = tempDeferredArray;
    }
    
    func sendMessage(message:CPMessage)
    {
        self.mSmHandler.addOperationAtEndOfQueue(message)
    }
    
    func sendMessageFront(message:CPMessage) {
        
        self.mSmHandler.addOperationAtFrontOfQueue(message);
    }
    
    func deferredMessage(message:CPMessage){
        
        self.mSmHandler.deferredArray.append(message)
    }
    
    func addState(state:State,parentState:State?) {
        
        self.mSmHandler.addState(state, parentState: parentState);
    }
    
}


